packages/core/mesh/network-manager/src/transport/libdatachannel-transport.ts

```diff
@@ -0,0 +1,241 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { Duplex } from 'stream';
+
+import { Event, Trigger } from '@dxos/async';
+import { ErrorStream } from '@dxos/debug';
+import { log } from '@dxos/log';
+import { Signal } from '@dxos/protocols/proto/dxos/mesh/swarm';
+
+import { type PeerConnection } from './datachannel/rtc-peer-connection';
+import { Transport } from './transport';
+
+export type LibDataChannelTransportParams = {
+  initiator: boolean;
+  stream: NodeJS.ReadWriteStream;
+  webrtcConfig?: RTCConfiguration;
+  sendSignal: (signal: Signal) => Promise<void>;
+};
+
+const MAX_BUFFERED_AMOUNT = 64 * 1024;
+
+// https://viblast.com/blog/2015/2/5/webrtc-data-channel-message-size/
+const MAX_MESSAGE_SIZE = 64 * 1024;
+
+/**
+ * Transport
+ */
+export class LibDataChannelTransport implements Transport {
+  private _closed = false;
+  readonly closed = new Event();
+  private _connected = false;
+  readonly connected = new Event();
+  readonly errors = new ErrorStream();
+  private readonly _peer: Promise<PeerConnection>;
+  private _channel!: RTCDataChannel;
+  private _stream!: Duplex;
+
+  private readonly _readyForCandidates = new Trigger();
+  private _writeCallback: (() => void) | null = null;
+
+  constructor(private readonly params: LibDataChannelTransportParams) {
+    this._peer = (async () => {
+      const { RTCPeerConnection: PeerConnection } = await import('./datachannel');
+      if (this._closed) {
+        this.errors.raise(Error('connection already closed'));
+      }
+      const peer = new PeerConnection(params.webrtcConfig);
+
+      peer.onicecandidateerror = (event) => {
+        log.error('peer.onicecandidateerror', { event });
+      };
+
+      peer.onconnectionstatechange = (event) => {
+        log.debug('peer.onconnectionstatechange', {
+          event,
+          peerConnectionState: peer.connectionState,
+          transportConnectionState: this._connected,
+        });
+        // TODO(nf): throw error if datachannel does not connect after some time?
+      };
+
+      peer.onicecandidate = async (event) => {
+        log.debug('peer.onicecandidate', { event });
+        if (event.candidate) {
+          await params.sendSignal({
+            payload: {
+              data: {
+                type: 'candidate',
+                candidate: {
+                  candidate: event.candidate.candidate,
+                  // these fields never seem to be not null, but connecting to Chrome doesn't work if they are
+                  sdpMLineIndex: event.candidate.sdpMLineIndex ?? 0,
+                  sdpMid: event.candidate.sdpMid ?? 0,
+                },
+              },
+            },
+          });
```

Comment:
This my error which would likely terminate the NodeJS process since the error happened in an event handler. For now we could just catch and log it since one candidate not being delivered is likely non-critical.

```suggestion
try {
          await params.sendSignal({
            payload: {
              data: {
                type: 'candidate',
                candidate: {
                  candidate: event.candidate.candidate,
                  // these fields never seem to be not null, but connecting to Chrome doesn't work if they are
                  sdpMLineIndex: event.candidate.sdpMLineIndex ?? 0,
                  sdpMid: event.candidate.sdpMid ?? 0,
                },
              },
            },
          });
          } catch(err) {
             log.warn('signaling errror, { err })
          }
```